daml-script/test/daml/upgrades/UpgradesTest.daml

```diff
@@ -0,0 +1,593 @@
+-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
+-- IMPORTANT: We do not track vetting state in these tests, it is your responsibility as the test writer to re-vet all packages you unvetted as part of your test at the end
+module UpgradesTest where
+
+import Daml.Script
+import qualified V1.MyTemplates as V1
+import qualified V2.MyTemplates as V2
+import V1.MyTemplates hiding (ValidUpgrade, SameTypeChoice)
+import V2.MyTemplates hiding (ValidUpgrade, SameTypeChoice)
+import DA.Foldable
+import DA.Text
+import DA.Assert
+import Daml.Script.Questions.Upgrading.Internal
+
+v1DarName : Text
+v1DarName = "upgrades-my-templates-v1"
+
+v2DarName : Text
+v2DarName = "upgrades-my-templates-v2"
+
+participant0 : ParticipantName
+participant0 = ParticipantName "participant0"
+
+participant1 : ParticipantName
+participant1 = ParticipantName "participant1"
+
+-- The error given back for unknown templates is either PACKAGE_NOT_FOUND (if the dar was never uploaded)
+-- or NO_DOMAIN_FOR_SUBMISSION. We always upload all dars before the test, so we'll never hit the former error.
+-- We've not yet implemented variants in the SubmitError type for NoDomainForSubmission, so we capture this error via UnknownError
+assertUnknownPackageError : Either SubmitError a -> Script ()
+assertUnknownPackageError (Left (UnknownError msg)) | "NO_DOMAIN_FOR_SUBMISSION" `isInfixOf` msg = pure ()
+assertUnknownPackageError (Left err) = assertFail $ "Expected NO_DOMAIN_FOR_SUBMISSION error, but got: " <> show err
+assertUnknownPackageError (Right _) = assertFail $ "Expected missing package error, but submission succeeded"
+
+-- Convenient wrapper we can inline with the submit call
+liftAssertUnknownPackageError : Script (Either SubmitError a) -> Script ()
+liftAssertUnknownPackageError s = s >>= assertUnknownPackageError
+
+-- Any queries or submits run within this block will have the package-id present.
+withExplicitPackageIds : Script a -> Script a
+withExplicitPackageIds act = do
+  setProvidePackageId True
+  ret <- act
+  setProvidePackageId False
+  pure ret
+
+main : Script ()
+main = forA_ tests $ \(testName, test) -> do
+  debugRaw $ "Testing: " <> testName
+  test
+
+tests : [(Text, Script ())]
+tests = mconcat
```

Comment:
Note. I would be nice to have something in Daml script to handle this kind of report in a non ad-hoc way.

Reply:
Agree, I'd like to implement something like this in future